### PR TITLE
Add cleanup for VPC load balancers

### PR DESCRIPTION
**Pull request by Amulyam24**
Add cleanup of load balancers for VPC resources

### CODE REVIEW

File: internal/ibmcloud-janitor/resources/types.go
Comment (mkumatag) [inline internal/ibmcloud-janitor/resources/types.go:50]: check the order, I believe loadbalancers will have subnets and need to be deleted before cleaning up the networks

need your help to review this PR further @dharaneeshvrd 
/cc @Prajyot-Parab

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (mkumatag) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:36]: see if this accepts the resource group or any other options to list whatever needs to be deleted instead of listing entire thing.

File: go.mod
Comment (mkumatag) [inline go.mod:9]: this update can go in a different PR to avoid dependency on the other maintainer for approving the code.

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go:36]: For options, we only have these
```
type ListLoadBalancersOptions struct {
    // A server-provided token determining what resource to start the page on.
    Start *string `json:"start,omitempty"`

    // The number of resources to return on a page.
    Limit *int64 `json:"limit,omitempty"`

    // Allows users to set headers on API requests
    Headers map[string]string
}
```

Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go:36]: Initial plan was to check with resource group similar to deleting instances & networks, but it isn't available

Comment (dharaneeshvrd) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go:36]: Yes listing load balancer is not providing any filters.

If the intention is to delete all the load balancers, then need to take care of paging as well.
https://github.com/openshift/hypershift/blob/main/cmd/infra/powervs/destroy.go#L735

File: internal/ibmcloud-janitor/resources/types.go
Comment (dharaneeshvrd) [reply internal/ibmcloud-janitor/resources/types.go]: Under subnets we will have load balancer as an attached resource, need to cleanup load balancer, before proceeding the deletion of subnet and then vpc.

Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/types.go]: Changed it.

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: Updated with changes, PTAL.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:53]: I think need to compare dereferenced value, not the pointer.

Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: That's right, fixed it.

**Comment by dharaneeshvrd** [comment]
In case of subnet deletion is attempted immediately after this cleanup, need to ensure, load balancer is deleted fully, before proceeding to the subnet or else subnet deletion will fail.
Like this block https://github.com/openshift/hypershift/blob/main/cmd/infra/powervs/destroy.go#L689

**Review by mkumatag — CHANGES_REQUESTED** [review]

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:73]: IMO we can split the deletion and checkLb separately, if we do like this sequentially it is going to take lot of time if we need to delete more LBs.
 
You can trigger the deletion on all the load balancers and process their deletion separately by storing deleted Lb IDs in a list or you can process this block in a goroutine maybe, so it would make things faster. I would prefer the 1st approach.

**Comment by Prajyot-Parab** [comment]
lgtm @dharaneeshvrd can you PTAL once.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:69]: I think this line should be inside if.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:43]: can we rename it to denote that it contains list of deleted LBs?

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:110]: Is it ok to use fmt.Println instead of resourceLogger.Info here?

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:97]: Is this required?

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go:43]: It's not initialised, need to use make.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:70]: can we log as "deletion triggered"? As at this point of time it won't be deleted fully.

Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: Changed it.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: Changed it.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: yes, fixed it.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go:99]: I added this because when I directly assign the id as `GetLoadBalancer(&vpcv1.GetLoadBalancerOptions{ID: &id}`, it was failing with a lint error.

**Comment by Amulyam24** [comment]
Review comments addressed, PTAL.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:72]: This line should also be inside if or else it would print for all the LB even if deletion not attempted.

Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: yes, missed it. Thanks for pointing it out.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:104]: This continue is not required

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:112]: Need to have a check here for to validate the LBs are processed or not before sleeping. 
Even though all the LB's deleted, it would sleep and loop till timeout becomes greater than lbTimeout.
Need to refactor this logic.

It's better if we can add some unit test cases.

Comment (dharaneeshvrd) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go:111]: @Amulyam24 We can use https://pkg.go.dev/k8s.io/apimachinery/pkg/util/wait@v0.25.2#PollImmediate this func to query something on a interval with a timeout.

I think this suits our requirement. Can you please take a look at this?

e.g usage https://github.com/openshift/hypershift/blob/main/cmd/infra/powervs/create.go#L727

Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: @dharaneeshvrd, made changes as per your suggestion. Tested it locally too, and it works fine. The LBs deletion took a little less than 2 minutes. I have added a polling interval of 30 secs and a timeout as 4 minutes. PTAL.

Comment (dharaneeshvrd) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: Ok the new logic looks good to me.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:69]: Do we need to come out of the loop before attempting to delete the remaining load balancers since one load balancer's delete API thrown an error?
IMO since it involves cleaning up all the load balancers, I would suggest to attempt the remaining load balancer deletion. And errors can be logged.
Same applies while monitoring the deletion as well, you can loop till end and any load balancer got an error while monitoring it can be logged.

Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: Addressed it, PTAL.

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:94]:
```suggestion
	checkLBs(deletedLBList, client, errs)
```

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:104]:
```suggestion
func checkLBs(list []string, client *IBMVPCClient, errs *[]error) {
```

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:122]:
```suggestion
```

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [inline internal/ibmcloud-janitor/resources/vpc_lbs.go:118]:
```suggestion
			*errs = append(*errs, err)
```

File: internal/ibmcloud-janitor/resources/vpc_lbs.go
Comment (dharaneeshvrd) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go:94]: If we don't pass by ref, then it's kind of appending errs with aggregated errs.

Comment (Amulyam24) [reply internal/ibmcloud-janitor/resources/vpc_lbs.go]: Addressed it. PTAL

**Comment by dharaneeshvrd** [comment]
LGTM
PTAL @mkumatag

**Review by mkumatag — APPROVED** [review]
/lgtm

**Pull request merged**
